src/main/java/org/jabref/gui/journals/AbbreviateAction.java

```diff
@@ -1,92 +1,179 @@
 package org.jabref.gui.journals;
 
-import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
-import java.util.concurrent.Callable;
-import java.util.concurrent.ExecutionException;
-import java.util.concurrent.Future;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
 
 import org.jabref.Globals;
-import org.jabref.JabRefExecutorService;
-import org.jabref.gui.BasePanel;
 import org.jabref.gui.actions.BaseAction;
 import org.jabref.gui.undo.NamedCompound;
 import org.jabref.gui.util.BackgroundTask;
-import org.jabref.logic.l10n.Localization;
+import org.jabref.model.database.BibDatabase;
 import org.jabref.model.entry.BibEntry;
 import org.jabref.model.entry.field.Field;
 import org.jabref.model.entry.field.FieldFactory;
 
+import org.reactfx.util.Lists;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /**
- * Converts journal full names to either iso or medline abbreviations for all
- * selected entries.
+ * Converts journal full names to either iso or medline abbreviations for all selected entries.
  */
 public class AbbreviateAction implements BaseAction {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(AbbreviateAction.class);
-    private final BasePanel panel;
-    private final boolean iso;
+    private final Supplier<List<BibEntry>> bibEntryListSupplier;
+    private final Supplier<UndoableAbbreviator> undoableAbbreviatorSupplier;
+    private final Supplier<BibDatabase> databaseSupplier;
+    private final Supplier<NamedCompound> compoundSupplier;
+    private final AbbreviateActionListener listener;
 
-    public AbbreviateAction(BasePanel panel, boolean iso) {
-        this.panel = panel;
-        this.iso = iso;
+    public AbbreviateAction(Supplier<List<BibEntry>> bibEntryListSupplier, Supplier<UndoableAbbreviator> undoableAbbreviatorSupplier, Supplier<BibDatabase> databaseSupplier, Supplier<NamedCompound> compoundSupplier, AbbreviateActionListener listener) {
```

Comment:
The point of these action classes is that they represent a single user interaction. Thus, there should be no (almost no) outside configuration possible. For this reason, I'm not a big fan of your changes.
For example, designing wise it's not bad that the AbbreviateAction handles logging and submitting of changes (although DialogService and UndoManager should be used directly insdead of BasePanel).

Reply:
Most of the 'outside configuration' has been passed into the class via the `BasePanel` object previously. This field dependecy to require a certain part of the GUI when you're actually processing data is completely dubious in my opinion. Therefore I simply removed this GUI related requirement, which makes this class a lot more flexible but still doing pretty much the same job as before.

Plase keep in mind not to 'weave' GUI components and data processing components into each other as this is really bad practice from an architectural point of view. But this is exactly what happenend in this case.

Reply:
I agree that there should be a clear separation of logic and gui, and most of JabRef's code sucks in this regard (although its getting better). However, the abbreviation class is not an example in my opinion. In contrast, it's actually quite good. Its prepares the input (get selected entries), invokes the logic class for the main action (UndoableAbbreviator) and then cleans up (submit changes to undo manager). That's relatively clean and straightforward (you could argue that the line `for (Field journalField : FieldFactory.getJournalNameFields()) {` should be moved to UndoableAbbreviator as it is business logic). You have created a flexible interface around the already pretty flexible UndoableAbbreviator - and I don't see the need for this honestly.  

One should keep in mind that the AbbreviateAction is a GUI class and thus should handle the UI interaction part of the feature. The base panel is definitely not the right place to do this as it is too big even now.

Reply:
> Its prepares the input (get selected entries), invokes the logic class for the main action (UndoableAbbreviator) and then cleans up (submit changes to undo manager).

Everything you described here is a clear indication of a data processing object and has to be strictly separated from any GUI related stuff. Therefore this class cannot be part of the GUI.
Even the JavaDoc of the class itself clearly indicates the manipulation of data which is **not** supposed to happen inside a GUI related object at all. 
`Converts journal full names to either iso or medline abbreviations for all selected entries.`

Following this model shown below, there has to be a strict separation of the three main layers 
(data, controller, UI) to provide a clean software architecture. (as we agreed already)
http://web.mit.edu/6.813/www/sp18/classes/05-ui-sw-arch/figures/05.png

These layers are supposed to communicate via interfaces to exchange data/information as needed.
I'm afraid to say, that there is currently a lot of data management/processing going on within the `BasePanel` class, which should be part of the data layer instead of the GUI layer.
Have a look at `BibDatabaseContext` for example. Usually a GUI does not need such information to work. Instead this class is used to initialize even more data related stuff like `SuggestionProviders` and `MainTableDataModel`

I'm very sorry to say, but in conclusion there is no doubt, that there is currently a huge mess going on in these GUI classes. And that's exactly what I was trying to achieve with this pull request - decouple data processing from GUI.

Best way to exchange the required data between data layer and GUI layer would be to create an interface that allows the `BasePanel` to retrieve exactly the data it requires or another interface to set the data into the GUI as it's available (that's by the way exactly what I programmed in this PR. Have a look at `AbbreviateActionListener`)

This way we'll achieve clean code and best maintainability for the whole project.

Reply:
In general, I totally agree. I think you just misunderstood how this class fits into JabRef's general architecture. Roughly speaking, data = model package, logic = logic package and gui = gui package. There does not need to be a further separation of "logic" and "data" in the gui as you cannot separate user input ("data") from initial pre-progressing ("logic") in a meaningful way.

If you want to see a real bad example: https://github.com/JabRef/jabref/blob/master/src/main/java/org/jabref/gui/importer/actions/OpenDatabaseAction.java
There the code is doing a lot of actual business logic that should be moved to the logic package. (But please don't start to refactor this code, it's so integral to JabRef that we do not want to touch it before the 5.0 release.)